concept/thing/thing_manager.rs

```diff
@@ -806,31 +838,226 @@ impl ThingManager {
+                    None => Err(ConceptReadError::CannotGetPlaysDoesntExist(
+                        owner_type.get_label(snapshot, self.type_manager())?.clone(),
+                        role_type.get_label(snapshot, self.type_manager())?.clone(),
+                    )),
+                    Some(plays) => Ok(plays),
+                }?
+                .into_owned();
+                self.add_exclusive_lock_for_plays_cardinality_constraint(snapshot, &player, plays)?;
+
+                let relates = match relation_type.get_relates_role(snapshot, self.type_manager(), role_type.clone())? {
+                    None => Err(ConceptReadError::CannotGetPlaysDoesntExist(
+                        owner_type.get_label(snapshot, self.type_manager())?.clone(),
+                        role_type.get_label(snapshot, self.type_manager())?.clone(),
+                    )),
+                    Some(plays) => Ok(plays),
+                }?
+                .into_owned();
+                self.add_exclusive_lock_for_relates_cardinality_constraint(snapshot, &relation, relates)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn add_exclusive_lock_for_unique_constraint<'a>(
+        &self,
+        snapshot: &mut impl WritableSnapshot,
+        owner: &Object<'a>,
+        value: Value<'a>,
+        owns: Owns<'static>,
+    ) -> Result<(), ConceptReadError> {
+        if let Some(uniqueness_source) = get_uniqueness_source(snapshot, self, owns)? {
+            let lock_key = create_custom_lock_key(
+                [
+                    &Infix::PropertyAnnotationUnique.infix_id().bytes(),
+                    uniqueness_source.attribute().vertex().bytes().bytes(),
+                    value.encode_bytes::<128>().bytes(), // TODO: Where to get 128 from?
+                    owner.vertex().bytes().bytes(),
+                    uniqueness_source.owner().vertex().bytes().bytes(),
+                ]
+                .into_iter(),
+            );
+            snapshot.exclusive_lock_add(lock_key);
+        }
+        Ok(())
+    }
+
+    fn add_exclusive_lock_for_owns_cardinality_constraint<'a>(
+        &self,
+        snapshot: &mut impl WritableSnapshot,
+        owner: &Object<'a>,
+        owns: Owns<'static>,
+    ) -> Result<(), ConceptReadError> {
+        let mut current_capability = Some(owns);
+        while let Some(locked_capability) = current_capability {
+            let cardinality = locked_capability.get_cardinality(snapshot, self.type_manager())?;
+            if cardinality == AnnotationCardinality::unchecked() {
+                break;
+            }
+
+            let lock_key = create_custom_lock_key(
+                [
+                    &Infix::PropertyAnnotationCardinality.infix_id().bytes(),
+                    owner.vertex().bytes().bytes(),
+                    locked_capability.interface().vertex().bytes().bytes(),
+                    // TODO: Do we need attribute type id?
```

Comment:
Attribute type id was mentioned in my notes, but I do believe that we already have enough for `card` locks here:
`Infix::AnnootationCardinality` + `owner's object` + `interface_type that is involved in the card check`

Reply:
I think we need the attribute type ID because you can have different card annotations for different ownerships of the object:
```
define
person owns id @card(1..1), owns email @card(0..10)
```
When you update the person's ID you need to lock `[@card][person1][has][id]` and [@card][person1][has][email]` otherwise transactions that dont actually overlap could affect each other with the lock `[@card][person1][has]`

Reply:
@flyingsilverfin Right now I use:
`[@card][person][id]` and `[@card][person][email]`. I don't think we need to specify `has` here because we don't have (and more likely won't have) situations when there is more than one capability between the same pair of `{object_type, interface_type}`.

Reply:
That is true, I think we can get away without it, there is never an overlap? Owns is always an attribute type, and therefore can never overlap with Plays cardinality, since that's always RoleType. The only thing that might overlap is `[card][Relation][Plays | Relates][RoleType]` right? If this is a possible overlap, I think we might want to be consistent and add the `has/plays/relates` all the time IMO

Reply:
`Relation` <-> `RoleType` cannot have both `Plays` and `Relates` with each other (we can't play what we relate), so I don't see a reason to. But if you want to super safe, I can add it.

Reply:
Actually, we can play what we relate (in a different relation instance!)
```
define
relation linked-list, relates previous, plays linked-list:previous;
```

Reply:
I'm actually not sure we have a reason to ban the same relation playing a role in itself either :P

Reply:
I never thought about it, what a monster :'(
Okay, I am convinced now

Reply:
Done